### DISCUSSION TITLE
Big Json Help

### DISCUSSION THREAD

**Discussion by lazyclaw-labs** [linq2db/linq2db#4738] [General]
I need help to save/read big json from text column.
For performance reason, I don't want to read everything into a string and deserialize to my POCO.
How to configure the mapping so I don't incur large objects (large strings) along the way of retrieval/insert? My Json could be over 100K, which is a big POCO.

I am thinking to use stream to convert to/from my POCO, but is it possible to plugin a stream converter in the process?

**Comment by lazyclaw-labs** [comment]
I guess what I really want is to be able to create a ValueConverter<T, Stream> for one of my POCO column, i.e., I can convert my POCO to/from stream, and linqtodb will happy accept a stream as a parameter and load into a stream and let me do the conversion on certain column.

**Comment by lazyclaw-labs** [comment]
The reason I have this request is:
Recently, my Dotnet app generate error message of OOM even I have plenty of RAM available, and it was determined that the LOH on the heap is the culprit. So I am working on minimize the Large Object in my App.
The reason of Large object is I do a lot of ToJson, and also save the json to database. ToJson create a big string, easily go over 100K, and 80K is the minimum size requirement of Large Object in dotnet.
I have eliminate LargeObject almost everywhere in my App except LinqToDB. Because it does not allow me to use stream. If I can use stream,  I can Serialize json into stream, and save to DB.

Right now, I have an alternative:  Serialize json into recyclablememorystream, save it to DB, dispose the buffer.

But I hit a wall on reading side. LinqToDB insist to read the text back to me before I can deserialize. They are big strings, certainly won't be garbage collectible easily.

I want a stream. Or Somehow allow me to give you a stream, or a buffer, so I can manage the memory myself. If you allocate a big string or a big byte[], if it's over 80K, I am at the mercy of garbage collector, and my app will crash at an undefined time, which is not good.

**Reply by MaceWindu** [reply]
Currently we don't have direct support for streams, but you probably could craft something depending on database provider you use (if it provides streaming API).

Converters alone probably will not be enough. For read operations you will need to subclass (or configure) linq2db database provider (only once at startup) to teach it how to use `GetStream` data reader methods by adding required read expressions to this collection https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/DataProvider/DataProviderBase.cs#L192

For write operations it's also depends on ADO.NET provider APIs. If it supports streams as parameters - you can use converter. If it provides custom APIs to setup stream as parameter, you probably also will need to subclass linq2db provider and override parameter setup methods

**Reply by lazyclaw-labs** [reply]
Thanks for the direction.
I can live without read operations for now. Can you give me a little bit more direction for write?
I am using postgresql, I am sure it supports stream parameter. Can you show me how to setup the converter?

subclass the provider is easy, but I can never get my head around the expressions. :-) And I am not sure how to configure to use my own provider. I will look into it though.

On the side note, I found it's hard to determine if a query is started, and done running. Not sure where to look. I have some special converters, for which I want to do some prep work when the query begin, and some cleanup work when query done running.

Thanks.

**Reply by MaceWindu** [reply]
For parameters you define converter from your type (Stream?) to DataParameter type in mapping schema. If Npgsql will not complain about parameter value being Stream - it's enough

**Reply by jods4** [reply]
Not sure how it works in Npgsql but with Oracle I've defined properties of type `OracleBlob` in my model and that was enough to be to read/write large data as streaming. 

Would be nice if linq2db supported `Stream` as a provider-agnostic way to do that.

**Reply by MaceWindu** [reply]
Yes, I've created tests for https://github.com/linq2db/linq2db/issues/1918 to see how it works currently  and it looks like this :)

```cs
[ActiveIssue(Configurations = [
	// cannot create table, DataType.Blob not mapped
	TestProvName.AllAccess, TestProvName.AllClickHouse, TestProvName.AllPostgreSQL, ProviderName.SqlCe,
	TestProvName.AllSqlServer, TestProvName.AllSybase,
	// fails on insert
	TestProvName.AllSQLite, TestProvName.AllSapHana, TestProvName.AllOracle, TestProvName.AllInformix, TestProvName.AllFirebird,
	// fails on select
	TestProvName.AllMySql, ProviderName.DB2
	])]
```

**Reply by lazyclaw-labs** [reply]
> For parameters you define converter from your type (Stream?) to DataParameter type in mapping schema. If Npgsql will not complain about parameter value being Stream - it's enough

Thanks. So for reader side, I just need to subclass or configure (Will figure out if it's possible to configure readerexpression) the provider, then I have a converter from stream to my type, it will work?

**Reply by MaceWindu** [reply]
Could you provide details which type your column has in database and to what type you want to map it. Like some custom type, stream or whatever

**Reply by lazyclaw-labs** [reply]
I have tried the DataParameter approach on the write side, it only works with bytea, not with text. Got error: 
 System.InvalidCastException: Can't write CLR type Microsoft.IO.RecyclableMemoryStream with handler type TextHandler

Somehow I can't find this TextHandler in NPgsql source code.

**Reply by MaceWindu** [reply]
Because if it is JSON(B) type mapped to custom DTO then I think you can almost use it as is with Npgsql. I did test JSON mapping for it recently and:
- provider itself handles poco de/serializing if you enable `EnableDynamicJson` option: https://www.npgsql.org/doc/types/json.html?tabs=datasource
- linq2db pass value to provider but cannot read it currently. For that you need to subclass provider and implement `GetFieldValue<T>` expression generation for such columns

**Reply by lazyclaw-labs** [reply]
I guess NPgsql does not use Stream at all. When it has a stream, it simply reads it into a byte[], it's my guess though.

OK, I have a text column, I want to map it to a POCO. from/to json automatically. It worked fine, but now I want to use stream, or at least use RecyclableMemoryStream so I can manage memory myself.  I probably could live with a bytea column, though text/json would be better.

Actually I already have write part done by pass a byte[] to the converter, it worked with Text column or bytea column. Stream only works with bytea column.

I can live with byte[] though stream could be better. since could manage memory for byte[].
-----------------
On the read side, still working on it.
I just don't want LinqToDB or NPgsql to allocate a big string or a big byte[], because I can't control. If I could provide the buffer byte[] myself, then I can manage the memory by using RecyclableMemoryStream. So I actually could live with with byte[] or stream.

It's just when reading, it seems I can't do either, for now. :-)

**Reply by lazyclaw-labs** [reply]
> Because if it is JSON(B) type mapped to custom DTO then I think you can almost use it as is with Npgsql. I did test JSON mapping for it recently and:
> 
> * provider itself handles poco de/serializing if you enable `EnableDynamicJson` option: https://www.npgsql.org/doc/types/json.html?tabs=datasource
> * linq2db pass value to provider but cannot read it currently. For that you need to subclass provider and implement `GetFieldValue<T>` expression generation for such columns

Let me test Json column. Never tried that.

**Reply by lazyclaw-labs** [reply]
Tried the Json/JsonB stuff, can not get the EnableDynamicJson() working, I called it, but it seems not effective.
But that's easy for me to figure it out later.
I am just not sure its implementation. If it does a simple toJson, then save the string, I won't be able to use it. I will find out.

**Reply by MaceWindu** [reply]
To use it with linq2db you need to register configured data source as connection factory
```cs
var builder = new NpgsqlDataSourceBuilder(connectionString);
builder.EnableDynamicJson();
var dataSource = builder.Build();

o.UseConnectionFactory(provider, _ => dataSource.CreateConnection());
```

**Reply by lazyclaw-labs** [reply]
Still can not get it work.
Not sure what I miss.
Never used datasource before. I am using DataContext to do the job. It keeps complaining I need to EnableDynamicJson() even I did, tried it in different places, and I debugged, saw the above code run.

Do I insert the above code in the datacontext  constructor? My DataContext subclass DataContext.

**Reply by MaceWindu** [reply]
Please provide sample code you try to make work - will check it later

**Reply by lazyclaw-labs** [reply]
My code is in a big project, may need to create a new project only test this.
In your code, where is the provider coming from?

var builder = new NpgsqlDataSourceBuilder(connectionString);
builder.EnableDynamicJson();
var dataSource = builder.Build();

o.UseConnectionFactory(provider, _ => dataSource.CreateConnection());

Here is a sample of my code:
 public class PostgreSqlDataContext : DataContext
 {
   static DataOptions createoption(LinqToDB.Configuration.ConnectionStringSettings css)
   {
     var connstr = (css.Name + "_CONNECTIONSTRING").AppSetting(css.ConnectionString);
     DataOptions o = new(new(ProviderName: css.ProviderName, ConnectionString: connstr) { ConfigurationString = css.Name });

     var builder = new NpgsqlDataSourceBuilder();
     builder.EnableDynamicJson();
     var ds = builder.Build();
     o.UseConnectionFactory( _ => ds.CreateConnection());
     return o;
   }

PostgreSqlDataContext (): base( createoption(Configuration.GetConnection("Default")))
{

}
}

**Reply by lazyclaw-labs** [reply]
o.UseConnectionFactory( _ => ds.CreateConnection());
Don't know where to get provider when creating options.
Right now, it's acting like dynamicjson not enabled at all.

**Reply by lazyclaw-labs** [reply]
If I use the old method,  NpgsqlConnection.GlobalTypeMapper.EnableDynamicJson();
It will work, but only on the write side. Read side still won't work.

**Reply by lazyclaw-labs** [reply]
This is the error message! Drive me crazy.
Seems missing something when apply this thing to LinqToDB.

 System.InvalidCastException: Writing values of 'LLMServer.Data.ItemSettings' is not supported for parameters having NpgsqlDbType 'Jsonb'. ---> System.NotSupportedException: Type 'ItemSettings' required dynamic JSON serialization, which requires an explicit opt-in; call 'EnableDynamicJson' on 'NpgsqlDataSourceBuilder' or NpgsqlConnection.GlobalTypeMapper (see https://www.npgsql.org/doc/types/json.html and the 8.0 release notes for more details). Alternatively, if you meant to use Newtonsoft JSON.NET instead of System.Text.Json, call UseJsonNet() instead.

**Reply by lazyclaw-labs** [reply]
Ley's forget NpgSql, and focus on creating our own solution on LinqToDB.
I have created and registered my DataProvider, and it worked, to some Extent I have intercepted the stream in my ValueConverter.
But the stream is always empty. Not sure what I missed, anyway, here is my code:
``` c#
class MyDataProvider : PostgreSQLDataProvider
{
  public MyDataProvider() : base(ProviderName.PostgreSQL15, PostgreSQLVersion.v15)
  {
    ReaderExpressions[new LinqToDB.DataProvider.ReaderInfo { ToType = typeof(Stream) }] = (DbDataReader r, int i) => r.GetStream(i);
  }

  public override CommandBehavior GetCommandBehavior(CommandBehavior commandBehavior)
  {
    return CommandBehavior.SequentialAccess | CommandBehavior.Default | CommandBehavior.CloseConnection;
    //return base.GetCommandBehavior(commandBehavior | CommandBehavior.SequentialAccess);
  }
}

 public class PostgreSqlDataContext : MyDataContext
 {
   //static NpgsqlDataSource dataSource;
   static PostgreSqlDataContext()
   {
     DataConnection.AddDataProvider(ProviderName.PostgreSQL, new MyDataProvider());
     DataConnection.AddDataProvider(ProviderName.PostgreSQL15, new MyDataProvider());
   }
}

public class JsonBytesConverter<T> : IValueConverter
{
  public bool HandlesNulls => false;
  public LambdaExpression FromProviderExpression => (Stream s) => JsonValueFromProvider(s);
  public LambdaExpression ToProviderExpression => (T p) => ConverterToProvider(p);
  private static DataParameter ConverterToProvider(T entity)
  {
    var mem = CDUtility.GetRecyclableMemoryStream();
    MyDataContext.Resources.Value?.Add(mem);
    JsonSerializer.Serialize(mem, entity, Extensions.prettyjo);
    return new DataParameter { Value = mem };
  }
  private static T JsonValueFromProvider(Stream value)
  {
    var str = value.ReadString(); //It's always empty string, seems like the stream is always empty.
    return JsonSerializer.Deserialize<T>(str) ?? CreateDefault();
  }
  static T CreateDefault() => (T)(typeof(T).IsArray ? Array.CreateInstance(typeof(T).GetElementType()!, 0) : typeof(T).CreateInstance())!;
}
```

**Reply by lazyclaw-labs** [reply]
I think it's because LinqToDB preread the content already! Not sure how to disable it.

**Comment by MaceWindu** [comment]
I've made some tests. `EnableDynamicJson` didn't worked with connection factory as it looks like we have configuration issue with it in 5.4.1. If you don't pass connection string to options, it uses factory as expected.

Below is an example that implements streaming support for all 4 types: json, jsonb, text and binary.

```cs
using LinqToDB;
using LinqToDB.Common;
using LinqToDB.Data;
using LinqToDB.DataProvider;
using LinqToDB.DataProvider.PostgreSQL;
using LinqToDB.Mapping;
using Npgsql;
using System.Data.Common;
using System.Linq.Expressions;
using System.Reflection;
using System.Text.Json;

internal class Program
{
	static void Main(string[] args)
	{
		const string connectionString = "...";

		var builder = new NpgsqlDataSourceBuilder(connectionString);
		builder.EnableDynamicJson();
		var ds = builder.Build();

		using var db = new DataConnection(new DataOptions().UseConnectionFactory(CustomProvider.Instance, o => ds.CreateConnection()));

		using var tb = db.CreateTempTable<JsonTable>();

		var record = new JsonTable()
		{
			Id = 1,
			JsonField = new JsonValue() { Value = 2 },
			JsonbField = new JsonValue() { Value = 3 },
			TextField = new JsonValue() { Value = 4 },
			BinaryField = new JsonValue() { Value = 5 },
		};

		db.Insert(record);

		var rawRecord = db.GetTable<JsonTableRaw>().Single();

		var loadedRecord = tb.Single();
	}
}

public sealed class CustomProvider : PostgreSQLDataProvider
{
	public static readonly IDataProvider Instance = new CustomProvider();

	private CustomProvider() : base(PostgreSQLVersion.v15)
	{
	}

	private static readonly MethodInfo _getFieldValue = typeof(DbDataReader).GetMethod(nameof(DbDataReader.GetFieldValue), BindingFlags.Instance | BindingFlags.Public)!;
	private static readonly MethodInfo _getStream = typeof(DbDataReader).GetMethod(nameof(DbDataReader.GetStream), BindingFlags.Instance | BindingFlags.Public)!;
	private static readonly ParameterExpression _rd = Expression.Parameter(typeof(DbDataReader));
	private static readonly ParameterExpression _ordinal = Expression.Parameter(typeof(int));

	public override Expression GetReaderExpression(DbDataReader reader, int idx, Expression readerExpression, Type? toType)
	{
		var typeName = reader.GetDataTypeName(idx);

		if (typeName is "json" or "jsonb")
		{
			// return rd.GetFieldValue<ToType>(ordinal)
			return Expression.Lambda(
				Expression.Call(_rd, _getFieldValue.MakeGenericMethod(toType!), _ordinal),
				_rd, _ordinal);
		}
		else if (typeName is "text" or "bytea" && toType == typeof(Stream))
		{
			// return rd.Stream(ordinal)
			return Expression.Lambda(
				Expression.Call(_rd, _getStream, _ordinal),
				_rd, _ordinal);
		}

		return base.GetReaderExpression(reader, idx, readerExpression, toType);
	}
}

[Table(nameof(JsonTable))]
public sealed class JsonTableRaw
{
	[PrimaryKey] public int Id { get; set; }

	[Column] public string? JsonField { get; set; }
	[Column] public string? JsonbField { get; set; }
	[Column] public string? TextField { get; set; }
	[Column] public byte[]? BinaryField { get; set; }
}

public sealed class JsonTable
{
	[PrimaryKey] public int Id { get; set; }

	[Column(DataType = DataType.Json)] public JsonValue? JsonField { get; set; }
	[Column(DataType = DataType.BinaryJson)] public JsonValue? JsonbField { get; set; }
	[ValueConverter(ConverterType = typeof(ObjectToStreamConverter<JsonValue>))]
	[Column(DataType = DataType.Text)] public JsonValue? TextField { get; set; }
	[ValueConverter(ConverterType = typeof(ObjectToStreamConverter<JsonValue>))]
	[Column(DataType = DataType.VarBinary)] public JsonValue? BinaryField { get; set; }
}

public sealed class ObjectToStreamConverter<TModel> : ValueConverter<TModel, Stream>
{
	public ObjectToStreamConverter()
		: base(m => ConvertToDatabase(m), s => ConvertToClient(s), false)
	{
	}

	private static Stream ConvertToDatabase(TModel clientValue)
	{
		var ms = new MemoryStream();
		JsonSerializer.Serialize(ms, clientValue);
		ms.Position = 0;
		return ms;
	}

	private static TModel ConvertToClient(Stream json)
	{
		return JsonSerializer.Deserialize<TModel>(json)!;
	}
}

public sealed class JsonValue
{
	public int Value { get; set; }
}
```

**Reply by lazyclaw-labs** [reply]
Thanks, I got it work.
It seems I have to enable DynamicJson in order to use stream. Which is a little bit Odd, as I don't plan to use the builtin capability. But if I don't enabledynamicjson, My stream would be an empty stream, but the error message indicate the content has been retrieved and shown in the error message.

On a sidenode, I found on JsonB data, the result of the stream read has an extra byte in front of the json. Did you have the same problem? Which is ODD, as your example does not have this issue. I will investigate later.

**Reply by lazyclaw-labs** [reply]
Turns out this is a pretty serious problem. Now I can't get connectionstring from DataContext any more.
Maybe I am not supposed to when UseConnectionFactory. But this really sucks, I am thinking that even UseConnectionFactory, it still should use the connectionstring from DataOptions.

We can work around it, just don't feel good about it now. NpgSqlDataSource is weird, not sure what it is. As I don't need it at all. Now I have to use it, screwed up ConnectionString, just because I want to enable Dynamic Json?

On the side note, Is NpgSql trying to do LinqToDB's job? As I even see data mapping there. I thought it's just a low level driver.

> I've made some tests. `EnableDynamicJson` didn't worked with connection factory as it looks like we have configuration issue with it in 5.4.1. If you don't pass connection string to options, it uses factory as expected.

**Reply by lazyclaw-labs** [reply]
Is this issue resolved in new version?

> I've made some tests. `EnableDynamicJson` didn't worked with connection factory as it looks like we have configuration issue with it in 5.4.1. If you don't pass connection string to options, it uses factory as expected.
> 
> Below is an example that implements streaming support for all 4 types: json, jsonb, text and binary.
